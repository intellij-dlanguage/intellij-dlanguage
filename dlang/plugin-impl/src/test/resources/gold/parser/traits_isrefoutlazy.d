void fooref(ref int x)
{
    static assert(__traits(isRef, x));
    static assert(!__traits(isOut, x));
    static assert(!__traits(isLazy, x));
}

void fooout(out int x)
{
    static assert(!__traits(isRef, x));
    static assert(__traits(isOut, x));
    static assert(!__traits(isLazy, x));
}

void foolazy(lazy int x)
{
    static assert(!__traits(isRef, x));
    static assert(!__traits(isOut, x));
    static assert(__traits(isLazy, x));
}