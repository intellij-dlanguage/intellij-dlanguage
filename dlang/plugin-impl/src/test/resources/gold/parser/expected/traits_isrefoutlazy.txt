D Language File
  DLanguageDeclarationImpl(DECLARATION)
    FUNCTION_DECLARATION
      DLanguageTypeImpl(TYPE)
        DLanguageType_2Impl(TYPE_2)
          DLanguageBuiltinTypeImpl(BUILTIN_TYPE)
            PsiElement(DlangTokenType.void)('void')
      IDENTIFIER
        PsiElement(DlangTokenType.ID)('fooref')
      DLanguageParametersImpl(PARAMETERS)
        PsiElement(DlangTokenType.()('(')
        PARAMETER
          PsiElement(DlangTokenType.ref)('ref')
          DLanguageTypeImpl(TYPE)
            DLanguageType_2Impl(TYPE_2)
              DLanguageBuiltinTypeImpl(BUILTIN_TYPE)
                PsiElement(DlangTokenType.int)('int')
          IDENTIFIER
            PsiElement(DlangTokenType.ID)('x')
        PsiElement(DlangTokenType.))(')')
      DLanguageFunctionBodyImpl(FUNCTION_BODY)
        DLanguageSpecifiedFunctionBodyImpl(SPECIFIED_FUNCTION_BODY)
          DLanguageBlockStatementImpl(BLOCK_STATEMENT)
            PsiElement(DlangTokenType.{)('{')
            DLanguageDeclarationsAndStatementsImpl(DECLARATIONS_AND_STATEMENTS)
              DLanguageDeclarationOrStatementImpl(DECLARATION_OR_STATEMENT)
                DLanguageDeclarationImpl(DECLARATION)
                  DLanguageStaticAssertDeclarationImpl(STATIC_ASSERT_DECLARATION)
                    DLanguageStaticAssertDeclarationImpl(STATIC_ASSERT_DECLARATION)
                      DLanguageStaticAssertStatementImpl(STATIC_ASSERT_STATEMENT)
                        DLanguageStaticAssertStatementImpl(STATIC_ASSERT_STATEMENT)
                          PsiElement(DlangTokenType.static)('static')
                          DLanguageAssertExpressionImpl(ASSERT_EXPRESSION)
                            PsiElement(DlangTokenType.assert)('assert')
                            PsiElement(DlangTokenType.()('(')
                            DLanguageAssertArgumentsImpl(ASSERT_ARGUMENTS)
                              DLanguageAssignExpressionImpl(ASSIGN_EXPRESSION)
                                DLanguageCmpExpressionImpl(CMP_EXPRESSION)
                                  DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                                    DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                                      DLanguageTraitsExpressionImpl(TRAITS_EXPRESSION)
                                        PsiElement(DlangTokenType.__traits)('__traits')
                                        PsiElement(DlangTokenType.()('(')
                                        IDENTIFIER
                                          PsiElement(DlangTokenType.ID)('isRef')
                                        PsiElement(DlangTokenType.,)(',')
                                        DLanguageTemplateArgumentListImpl(TEMPLATE_ARGUMENT_LIST)
                                          DLanguageTemplateArgumentImpl(TEMPLATE_ARGUMENT)
                                            DLanguageTypeImpl(TYPE)
                                              DLanguageType_2Impl(TYPE_2)
                                                DLanguageTypeIdentifierPartImpl(TYPE_IDENTIFIER_PART)
                                                  DLanguageIdentifierOrTemplateInstanceImpl(IDENTIFIER_OR_TEMPLATE_INSTANCE)
                                                    IDENTIFIER
                                                      PsiElement(DlangTokenType.ID)('x')
                                        PsiElement(DlangTokenType.))(')')
                            PsiElement(DlangTokenType.))(')')
                          PsiElement(DlangTokenType.;)(';')
              DLanguageDeclarationOrStatementImpl(DECLARATION_OR_STATEMENT)
                DLanguageDeclarationImpl(DECLARATION)
                  DLanguageStaticAssertDeclarationImpl(STATIC_ASSERT_DECLARATION)
                    DLanguageStaticAssertDeclarationImpl(STATIC_ASSERT_DECLARATION)
                      DLanguageStaticAssertStatementImpl(STATIC_ASSERT_STATEMENT)
                        DLanguageStaticAssertStatementImpl(STATIC_ASSERT_STATEMENT)
                          PsiElement(DlangTokenType.static)('static')
                          DLanguageAssertExpressionImpl(ASSERT_EXPRESSION)
                            PsiElement(DlangTokenType.assert)('assert')
                            PsiElement(DlangTokenType.()('(')
                            DLanguageAssertArgumentsImpl(ASSERT_ARGUMENTS)
                              DLanguageAssignExpressionImpl(ASSIGN_EXPRESSION)
                                DLanguageCmpExpressionImpl(CMP_EXPRESSION)
                                  DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                                    PsiElement(DlangTokenType.!)('!')
                                    DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                                      DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                                        DLanguageTraitsExpressionImpl(TRAITS_EXPRESSION)
                                          PsiElement(DlangTokenType.__traits)('__traits')
                                          PsiElement(DlangTokenType.()('(')
                                          IDENTIFIER
                                            PsiElement(DlangTokenType.ID)('isOut')
                                          PsiElement(DlangTokenType.,)(',')
                                          DLanguageTemplateArgumentListImpl(TEMPLATE_ARGUMENT_LIST)
                                            DLanguageTemplateArgumentImpl(TEMPLATE_ARGUMENT)
                                              DLanguageTypeImpl(TYPE)
                                                DLanguageType_2Impl(TYPE_2)
                                                  DLanguageTypeIdentifierPartImpl(TYPE_IDENTIFIER_PART)
                                                    DLanguageIdentifierOrTemplateInstanceImpl(IDENTIFIER_OR_TEMPLATE_INSTANCE)
                                                      IDENTIFIER
                                                        PsiElement(DlangTokenType.ID)('x')
                                          PsiElement(DlangTokenType.))(')')
                            PsiElement(DlangTokenType.))(')')
                          PsiElement(DlangTokenType.;)(';')
              DLanguageDeclarationOrStatementImpl(DECLARATION_OR_STATEMENT)
                DLanguageDeclarationImpl(DECLARATION)
                  DLanguageStaticAssertDeclarationImpl(STATIC_ASSERT_DECLARATION)
                    DLanguageStaticAssertDeclarationImpl(STATIC_ASSERT_DECLARATION)
                      DLanguageStaticAssertStatementImpl(STATIC_ASSERT_STATEMENT)
                        DLanguageStaticAssertStatementImpl(STATIC_ASSERT_STATEMENT)
                          PsiElement(DlangTokenType.static)('static')
                          DLanguageAssertExpressionImpl(ASSERT_EXPRESSION)
                            PsiElement(DlangTokenType.assert)('assert')
                            PsiElement(DlangTokenType.()('(')
                            DLanguageAssertArgumentsImpl(ASSERT_ARGUMENTS)
                              DLanguageAssignExpressionImpl(ASSIGN_EXPRESSION)
                                DLanguageCmpExpressionImpl(CMP_EXPRESSION)
                                  DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                                    PsiElement(DlangTokenType.!)('!')
                                    DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                                      DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                                        DLanguageTraitsExpressionImpl(TRAITS_EXPRESSION)
                                          PsiElement(DlangTokenType.__traits)('__traits')
                                          PsiElement(DlangTokenType.()('(')
                                          IDENTIFIER
                                            PsiElement(DlangTokenType.ID)('isLazy')
                                          PsiElement(DlangTokenType.,)(',')
                                          DLanguageTemplateArgumentListImpl(TEMPLATE_ARGUMENT_LIST)
                                            DLanguageTemplateArgumentImpl(TEMPLATE_ARGUMENT)
                                              DLanguageTypeImpl(TYPE)
                                                DLanguageType_2Impl(TYPE_2)
                                                  DLanguageTypeIdentifierPartImpl(TYPE_IDENTIFIER_PART)
                                                    DLanguageIdentifierOrTemplateInstanceImpl(IDENTIFIER_OR_TEMPLATE_INSTANCE)
                                                      IDENTIFIER
                                                        PsiElement(DlangTokenType.ID)('x')
                                          PsiElement(DlangTokenType.))(')')
                            PsiElement(DlangTokenType.))(')')
                          PsiElement(DlangTokenType.;)(';')
            PsiElement(DlangTokenType.})('}')
  DLanguageDeclarationImpl(DECLARATION)
    FUNCTION_DECLARATION
      DLanguageTypeImpl(TYPE)
        DLanguageType_2Impl(TYPE_2)
          DLanguageBuiltinTypeImpl(BUILTIN_TYPE)
            PsiElement(DlangTokenType.void)('void')
      IDENTIFIER
        PsiElement(DlangTokenType.ID)('fooout')
      DLanguageParametersImpl(PARAMETERS)
        PsiElement(DlangTokenType.()('(')
        PARAMETER
          PsiElement(DlangTokenType.out)('out')
          DLanguageTypeImpl(TYPE)
            DLanguageType_2Impl(TYPE_2)
              DLanguageBuiltinTypeImpl(BUILTIN_TYPE)
                PsiElement(DlangTokenType.int)('int')
          IDENTIFIER
            PsiElement(DlangTokenType.ID)('x')
        PsiElement(DlangTokenType.))(')')
      DLanguageFunctionBodyImpl(FUNCTION_BODY)
        DLanguageSpecifiedFunctionBodyImpl(SPECIFIED_FUNCTION_BODY)
          DLanguageBlockStatementImpl(BLOCK_STATEMENT)
            PsiElement(DlangTokenType.{)('{')
            DLanguageDeclarationsAndStatementsImpl(DECLARATIONS_AND_STATEMENTS)
              DLanguageDeclarationOrStatementImpl(DECLARATION_OR_STATEMENT)
                DLanguageDeclarationImpl(DECLARATION)
                  DLanguageStaticAssertDeclarationImpl(STATIC_ASSERT_DECLARATION)
                    DLanguageStaticAssertDeclarationImpl(STATIC_ASSERT_DECLARATION)
                      DLanguageStaticAssertStatementImpl(STATIC_ASSERT_STATEMENT)
                        DLanguageStaticAssertStatementImpl(STATIC_ASSERT_STATEMENT)
                          PsiElement(DlangTokenType.static)('static')
                          DLanguageAssertExpressionImpl(ASSERT_EXPRESSION)
                            PsiElement(DlangTokenType.assert)('assert')
                            PsiElement(DlangTokenType.()('(')
                            DLanguageAssertArgumentsImpl(ASSERT_ARGUMENTS)
                              DLanguageAssignExpressionImpl(ASSIGN_EXPRESSION)
                                DLanguageCmpExpressionImpl(CMP_EXPRESSION)
                                  DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                                    PsiElement(DlangTokenType.!)('!')
                                    DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                                      DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                                        DLanguageTraitsExpressionImpl(TRAITS_EXPRESSION)
                                          PsiElement(DlangTokenType.__traits)('__traits')
                                          PsiElement(DlangTokenType.()('(')
                                          IDENTIFIER
                                            PsiElement(DlangTokenType.ID)('isRef')
                                          PsiElement(DlangTokenType.,)(',')
                                          DLanguageTemplateArgumentListImpl(TEMPLATE_ARGUMENT_LIST)
                                            DLanguageTemplateArgumentImpl(TEMPLATE_ARGUMENT)
                                              DLanguageTypeImpl(TYPE)
                                                DLanguageType_2Impl(TYPE_2)
                                                  DLanguageTypeIdentifierPartImpl(TYPE_IDENTIFIER_PART)
                                                    DLanguageIdentifierOrTemplateInstanceImpl(IDENTIFIER_OR_TEMPLATE_INSTANCE)
                                                      IDENTIFIER
                                                        PsiElement(DlangTokenType.ID)('x')
                                          PsiElement(DlangTokenType.))(')')
                            PsiElement(DlangTokenType.))(')')
                          PsiElement(DlangTokenType.;)(';')
              DLanguageDeclarationOrStatementImpl(DECLARATION_OR_STATEMENT)
                DLanguageDeclarationImpl(DECLARATION)
                  DLanguageStaticAssertDeclarationImpl(STATIC_ASSERT_DECLARATION)
                    DLanguageStaticAssertDeclarationImpl(STATIC_ASSERT_DECLARATION)
                      DLanguageStaticAssertStatementImpl(STATIC_ASSERT_STATEMENT)
                        DLanguageStaticAssertStatementImpl(STATIC_ASSERT_STATEMENT)
                          PsiElement(DlangTokenType.static)('static')
                          DLanguageAssertExpressionImpl(ASSERT_EXPRESSION)
                            PsiElement(DlangTokenType.assert)('assert')
                            PsiElement(DlangTokenType.()('(')
                            DLanguageAssertArgumentsImpl(ASSERT_ARGUMENTS)
                              DLanguageAssignExpressionImpl(ASSIGN_EXPRESSION)
                                DLanguageCmpExpressionImpl(CMP_EXPRESSION)
                                  DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                                    DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                                      DLanguageTraitsExpressionImpl(TRAITS_EXPRESSION)
                                        PsiElement(DlangTokenType.__traits)('__traits')
                                        PsiElement(DlangTokenType.()('(')
                                        IDENTIFIER
                                          PsiElement(DlangTokenType.ID)('isOut')
                                        PsiElement(DlangTokenType.,)(',')
                                        DLanguageTemplateArgumentListImpl(TEMPLATE_ARGUMENT_LIST)
                                          DLanguageTemplateArgumentImpl(TEMPLATE_ARGUMENT)
                                            DLanguageTypeImpl(TYPE)
                                              DLanguageType_2Impl(TYPE_2)
                                                DLanguageTypeIdentifierPartImpl(TYPE_IDENTIFIER_PART)
                                                  DLanguageIdentifierOrTemplateInstanceImpl(IDENTIFIER_OR_TEMPLATE_INSTANCE)
                                                    IDENTIFIER
                                                      PsiElement(DlangTokenType.ID)('x')
                                        PsiElement(DlangTokenType.))(')')
                            PsiElement(DlangTokenType.))(')')
                          PsiElement(DlangTokenType.;)(';')
              DLanguageDeclarationOrStatementImpl(DECLARATION_OR_STATEMENT)
                DLanguageDeclarationImpl(DECLARATION)
                  DLanguageStaticAssertDeclarationImpl(STATIC_ASSERT_DECLARATION)
                    DLanguageStaticAssertDeclarationImpl(STATIC_ASSERT_DECLARATION)
                      DLanguageStaticAssertStatementImpl(STATIC_ASSERT_STATEMENT)
                        DLanguageStaticAssertStatementImpl(STATIC_ASSERT_STATEMENT)
                          PsiElement(DlangTokenType.static)('static')
                          DLanguageAssertExpressionImpl(ASSERT_EXPRESSION)
                            PsiElement(DlangTokenType.assert)('assert')
                            PsiElement(DlangTokenType.()('(')
                            DLanguageAssertArgumentsImpl(ASSERT_ARGUMENTS)
                              DLanguageAssignExpressionImpl(ASSIGN_EXPRESSION)
                                DLanguageCmpExpressionImpl(CMP_EXPRESSION)
                                  DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                                    PsiElement(DlangTokenType.!)('!')
                                    DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                                      DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                                        DLanguageTraitsExpressionImpl(TRAITS_EXPRESSION)
                                          PsiElement(DlangTokenType.__traits)('__traits')
                                          PsiElement(DlangTokenType.()('(')
                                          IDENTIFIER
                                            PsiElement(DlangTokenType.ID)('isLazy')
                                          PsiElement(DlangTokenType.,)(',')
                                          DLanguageTemplateArgumentListImpl(TEMPLATE_ARGUMENT_LIST)
                                            DLanguageTemplateArgumentImpl(TEMPLATE_ARGUMENT)
                                              DLanguageTypeImpl(TYPE)
                                                DLanguageType_2Impl(TYPE_2)
                                                  DLanguageTypeIdentifierPartImpl(TYPE_IDENTIFIER_PART)
                                                    DLanguageIdentifierOrTemplateInstanceImpl(IDENTIFIER_OR_TEMPLATE_INSTANCE)
                                                      IDENTIFIER
                                                        PsiElement(DlangTokenType.ID)('x')
                                          PsiElement(DlangTokenType.))(')')
                            PsiElement(DlangTokenType.))(')')
                          PsiElement(DlangTokenType.;)(';')
            PsiElement(DlangTokenType.})('}')
  DLanguageDeclarationImpl(DECLARATION)
    FUNCTION_DECLARATION
      DLanguageTypeImpl(TYPE)
        DLanguageType_2Impl(TYPE_2)
          DLanguageBuiltinTypeImpl(BUILTIN_TYPE)
            PsiElement(DlangTokenType.void)('void')
      IDENTIFIER
        PsiElement(DlangTokenType.ID)('foolazy')
      DLanguageParametersImpl(PARAMETERS)
        PsiElement(DlangTokenType.()('(')
        PARAMETER
          PsiElement(DlangTokenType.lazy)('lazy')
          DLanguageTypeImpl(TYPE)
            DLanguageType_2Impl(TYPE_2)
              DLanguageBuiltinTypeImpl(BUILTIN_TYPE)
                PsiElement(DlangTokenType.int)('int')
          IDENTIFIER
            PsiElement(DlangTokenType.ID)('x')
        PsiElement(DlangTokenType.))(')')
      DLanguageFunctionBodyImpl(FUNCTION_BODY)
        DLanguageSpecifiedFunctionBodyImpl(SPECIFIED_FUNCTION_BODY)
          DLanguageBlockStatementImpl(BLOCK_STATEMENT)
            PsiElement(DlangTokenType.{)('{')
            DLanguageDeclarationsAndStatementsImpl(DECLARATIONS_AND_STATEMENTS)
              DLanguageDeclarationOrStatementImpl(DECLARATION_OR_STATEMENT)
                DLanguageDeclarationImpl(DECLARATION)
                  DLanguageStaticAssertDeclarationImpl(STATIC_ASSERT_DECLARATION)
                    DLanguageStaticAssertDeclarationImpl(STATIC_ASSERT_DECLARATION)
                      DLanguageStaticAssertStatementImpl(STATIC_ASSERT_STATEMENT)
                        DLanguageStaticAssertStatementImpl(STATIC_ASSERT_STATEMENT)
                          PsiElement(DlangTokenType.static)('static')
                          DLanguageAssertExpressionImpl(ASSERT_EXPRESSION)
                            PsiElement(DlangTokenType.assert)('assert')
                            PsiElement(DlangTokenType.()('(')
                            DLanguageAssertArgumentsImpl(ASSERT_ARGUMENTS)
                              DLanguageAssignExpressionImpl(ASSIGN_EXPRESSION)
                                DLanguageCmpExpressionImpl(CMP_EXPRESSION)
                                  DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                                    PsiElement(DlangTokenType.!)('!')
                                    DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                                      DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                                        DLanguageTraitsExpressionImpl(TRAITS_EXPRESSION)
                                          PsiElement(DlangTokenType.__traits)('__traits')
                                          PsiElement(DlangTokenType.()('(')
                                          IDENTIFIER
                                            PsiElement(DlangTokenType.ID)('isRef')
                                          PsiElement(DlangTokenType.,)(',')
                                          DLanguageTemplateArgumentListImpl(TEMPLATE_ARGUMENT_LIST)
                                            DLanguageTemplateArgumentImpl(TEMPLATE_ARGUMENT)
                                              DLanguageTypeImpl(TYPE)
                                                DLanguageType_2Impl(TYPE_2)
                                                  DLanguageTypeIdentifierPartImpl(TYPE_IDENTIFIER_PART)
                                                    DLanguageIdentifierOrTemplateInstanceImpl(IDENTIFIER_OR_TEMPLATE_INSTANCE)
                                                      IDENTIFIER
                                                        PsiElement(DlangTokenType.ID)('x')
                                          PsiElement(DlangTokenType.))(')')
                            PsiElement(DlangTokenType.))(')')
                          PsiElement(DlangTokenType.;)(';')
              DLanguageDeclarationOrStatementImpl(DECLARATION_OR_STATEMENT)
                DLanguageDeclarationImpl(DECLARATION)
                  DLanguageStaticAssertDeclarationImpl(STATIC_ASSERT_DECLARATION)
                    DLanguageStaticAssertDeclarationImpl(STATIC_ASSERT_DECLARATION)
                      DLanguageStaticAssertStatementImpl(STATIC_ASSERT_STATEMENT)
                        DLanguageStaticAssertStatementImpl(STATIC_ASSERT_STATEMENT)
                          PsiElement(DlangTokenType.static)('static')
                          DLanguageAssertExpressionImpl(ASSERT_EXPRESSION)
                            PsiElement(DlangTokenType.assert)('assert')
                            PsiElement(DlangTokenType.()('(')
                            DLanguageAssertArgumentsImpl(ASSERT_ARGUMENTS)
                              DLanguageAssignExpressionImpl(ASSIGN_EXPRESSION)
                                DLanguageCmpExpressionImpl(CMP_EXPRESSION)
                                  DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                                    PsiElement(DlangTokenType.!)('!')
                                    DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                                      DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                                        DLanguageTraitsExpressionImpl(TRAITS_EXPRESSION)
                                          PsiElement(DlangTokenType.__traits)('__traits')
                                          PsiElement(DlangTokenType.()('(')
                                          IDENTIFIER
                                            PsiElement(DlangTokenType.ID)('isOut')
                                          PsiElement(DlangTokenType.,)(',')
                                          DLanguageTemplateArgumentListImpl(TEMPLATE_ARGUMENT_LIST)
                                            DLanguageTemplateArgumentImpl(TEMPLATE_ARGUMENT)
                                              DLanguageTypeImpl(TYPE)
                                                DLanguageType_2Impl(TYPE_2)
                                                  DLanguageTypeIdentifierPartImpl(TYPE_IDENTIFIER_PART)
                                                    DLanguageIdentifierOrTemplateInstanceImpl(IDENTIFIER_OR_TEMPLATE_INSTANCE)
                                                      IDENTIFIER
                                                        PsiElement(DlangTokenType.ID)('x')
                                          PsiElement(DlangTokenType.))(')')
                            PsiElement(DlangTokenType.))(')')
                          PsiElement(DlangTokenType.;)(';')
              DLanguageDeclarationOrStatementImpl(DECLARATION_OR_STATEMENT)
                DLanguageDeclarationImpl(DECLARATION)
                  DLanguageStaticAssertDeclarationImpl(STATIC_ASSERT_DECLARATION)
                    DLanguageStaticAssertDeclarationImpl(STATIC_ASSERT_DECLARATION)
                      DLanguageStaticAssertStatementImpl(STATIC_ASSERT_STATEMENT)
                        DLanguageStaticAssertStatementImpl(STATIC_ASSERT_STATEMENT)
                          PsiElement(DlangTokenType.static)('static')
                          DLanguageAssertExpressionImpl(ASSERT_EXPRESSION)
                            PsiElement(DlangTokenType.assert)('assert')
                            PsiElement(DlangTokenType.()('(')
                            DLanguageAssertArgumentsImpl(ASSERT_ARGUMENTS)
                              DLanguageAssignExpressionImpl(ASSIGN_EXPRESSION)
                                DLanguageCmpExpressionImpl(CMP_EXPRESSION)
                                  DLanguageUnaryExpressionImpl(UNARY_EXPRESSION)
                                    DLanguagePrimaryExpressionImpl(PRIMARY_EXPRESSION)
                                      DLanguageTraitsExpressionImpl(TRAITS_EXPRESSION)
                                        PsiElement(DlangTokenType.__traits)('__traits')
                                        PsiElement(DlangTokenType.()('(')
                                        IDENTIFIER
                                          PsiElement(DlangTokenType.ID)('isLazy')
                                        PsiElement(DlangTokenType.,)(',')
                                        DLanguageTemplateArgumentListImpl(TEMPLATE_ARGUMENT_LIST)
                                          DLanguageTemplateArgumentImpl(TEMPLATE_ARGUMENT)
                                            DLanguageTypeImpl(TYPE)
                                              DLanguageType_2Impl(TYPE_2)
                                                DLanguageTypeIdentifierPartImpl(TYPE_IDENTIFIER_PART)
                                                  DLanguageIdentifierOrTemplateInstanceImpl(IDENTIFIER_OR_TEMPLATE_INSTANCE)
                                                    IDENTIFIER
                                                      PsiElement(DlangTokenType.ID)('x')
                                        PsiElement(DlangTokenType.))(')')
                            PsiElement(DlangTokenType.))(')')
                          PsiElement(DlangTokenType.;)(';')
            PsiElement(DlangTokenType.})('}')